samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/CalculatedQuestionAddGlobalVariableListener.java

```diff
@@ -0,0 +1,138 @@
+     * <li>variables with invalid ranges of values</li>
+     * <li>formulas that are syntactically wrong</li></ul>
+     * Any errors are written to the context messager
+     * <p>The validate formula is also called directly from the ItemAddListner, before
+     * saving a calculated question, to ensure any last minute changes are caught.
+     */
+    public void processAction(ActionEvent arg0) throws AbortProcessingException {
+        ItemAuthorBean itemauthorbean = (ItemAuthorBean) ContextUtil.lookupBean("itemauthor");
+        ItemBean item = itemauthorbean.getCurrentItem();
+
+        FacesContext context = FacesContext.getCurrentInstance();
+        String globalvariablename = ContextUtil.lookupParam("globalvariablename");
+        List<String> errors = this.validate(item, globalvariablename, true);
+
+        if (errors.size() > 0) {
+            item.setOutcome("calculatedQuestion");
+            item.setPoolOutcome("calculatedQuestion");
+            for (String error : errors) {
+                context.addMessage(null, new FacesMessage(error));
+            }
+            context.renderResponse();
+        }
+    }
+
+    /**
+     * validate() returns a list of error strings to display to the context.
+     * <p>Errors include <ul><li>global variables is already included</li>
+     * @param item - an ItemBean, which contains all of the needed information 
+     * about the CalculatedQuestion
+     * @param globalvariablename
+     * @param fromui - boolean. Indicate if global variable has been added from ui.
+     * @returns a List<String> of error messages to be displayed in the context messager.
+     */
+    public List<String> validate(ItemBean item, String globalvariablename, boolean fromui) {
+        List<String> errors = new ArrayList<String>();
+
+        Map<String, CalculatedQuestionFormulaBean> formulas = item.getCalculatedQuestion().getFormulas();
+        Map<String, CalculatedQuestionVariableBean> variables = item.getCalculatedQuestion().getVariables();
+        Map<String, CalculatedQuestionGlobalVariableBean> globalVariables = item.getCalculatedQuestion().getGlobalvariables();
+        Long maxSequenceValue = getMaxSequenceValue(variables, formulas, globalVariables);
+
+        if (!globalVariables.containsKey(globalvariablename)) {
+            CalculatedQuestionGlobalVariableBean bean = new CalculatedQuestionGlobalVariableBean(globalvariablename);
+            bean.setName(globalvariablename);
+            bean.setFromui(fromui);
+            bean.setActive(true);
+            bean.setSequence(++maxSequenceValue);
+            item.getCalculatedQuestion().addGlobalVariable(bean);
+        } else {
+            errors.add(getErrorMessage("unique_names"));
+        }
+        return errors;
+    }
+
+    private Long getMaxSequenceValue(Map<String, CalculatedQuestionVariableBean> variables, Map<String, CalculatedQuestionFormulaBean> formulas, Map<String, CalculatedQuestionGlobalVariableBean> globalVariables) {

```

Comment:
brackets and alignment, for example

private Long getMaxSequenceValue(Map<String, CalculatedQuestionVariableBean> variables,
    Map<String, CalculatedQuestionFormulaBean> formulas,
    Map<String, CalculatedQuestionGlobalVariableBean> globalVariables ) {

Reply:
Changed